src/Compiler/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs

```diff
@@ -64,60 +130,104 @@ public string NewLine
     {
         get
         {
-            if (index < 0 || index >= _builder.Length)
+            // This Debug.Fail(...) is present because no Razor code currently accesses this
+            // indexer and it isn't implemented efficiently. All Razor code that previously
+            // used the indexer were really just inspecting the last char, which is now exposed separately.
+            Debug.Fail("Do not use this indexer without reimplementing it more efficiently.");
+
+            foreach (var page in _pages)
             {
-                throw new IndexOutOfRangeException(nameof(index));
+                foreach (var chars in page)
+                {
+                    if (index < chars.Length)
+                    {
+                        return chars.Span[index];
+                    }
+
+                    index -= chars.Length;
+                }
             }
 
-            return _builder[index];
+            throw new IndexOutOfRangeException(nameof(index));
         }
     }
 
+    public char? LastChar => _lastChar;
+
     public CodeWriter Indent(int size)
     {
-        if (size == 0 || (Length != 0 && this[Length - 1] != '\n'))
+        if (size == 0 || LastChar is not '\n')
         {
             return this;
         }
 
-        var actualSize = 0;
-        if (IndentWithTabs)
-        {
-            // Avoid writing directly to the StringBuilder here, that will throw off the manual indexing
-            // done by the base class.
-            var tabs = size / TabSize;
-            actualSize += tabs;
-            _builder.Append('\t', tabs);
+        var indentString = size == _indentSize
+            ? _indentString
+            : ComputeIndent(size, IndentWithTabs, TabSize);
 
-            var spaces = size % TabSize;
-            actualSize += spaces;
-            _builder.Append(' ', spaces);
-        }
-        else
+        AddTextChunk(indentString);
+
+        var indentLength = indentString.Length;
+        _currentLineCharacterIndex += indentLength;
+        _absoluteIndex += indentLength;
+
+        return this;
+    }
+
+    public CodeWriter Indent()
```

Comment:
Consider simply calling `Indent(_indentSize)` to avoid code duplication.

Reply:
Will do. This particular method looked a lot different in earlier iterations. Now that they're super-similar, there's no reason to keep both.

Reply:
It's still actually used in two places -- it's used in the IndentSize property setter. So, it can't be a local function.